airflow/www/async_dag_loaders.py

```diff
@@ -0,0 +1,202 @@
+            dagbag_update = (dagbag, previous_keys, collect_done)
+
+            if dagbag_update or collect_done:
+                queue.put((collect_done, dagbag_update))
+
+            if collect_done:
+                for v in dagbag.values():
+                    v.clear()
+
+                event_collect_done.clear()
+                event_next_collect.set()
+
+            time.sleep(DAGBAG_SYNC_INTERVAL)
+        except Exception:  # pylint: disable=broad-except
+            logging.warning('Dagbag loader sender errors.', exc_info=True)
+
+
+def _create_dagbag(dag_folder, queue):
+    """A process that creates, updates, and sync dagbag in background."""
+    import airflow
+    # pylint: disable=redefined-outer-name,reimported
+    from airflow import configuration
+
+    # TODO(coufon): env vars that are not exposed to subprocesses are invisible here.
+    # In Composer we reload all env vars here, should check whether a similar step may
+    # be needed.
+
+    configuration = importlib.reload(configuration)
+    airflow.configuration = importlib.reload(airflow.configuration)
+    airflow.plugins_manager = importlib.reload(airflow.plugins_manager)
+    airflow = importlib.reload(airflow)
+
+    signal.signal(signal.SIGTERM, _kill_proc)
+    signal.signal(signal.SIGINT, _kill_proc)
+
+    logging.info('Using Asynchronous Dagbag Loader.')
+    dagbag = DagBag(dag_folder)
+    event_collect_done = Event()
+    event_next_collect = Event()
+    thread = threading.Thread(target=_send_dagbag,
+                              args=(dagbag, queue, event_collect_done,
+                                    event_next_collect))
+    thread.daemon = True
+    thread.start()
+    while True:
+        try:
+            event_collect_done.clear()
+            event_next_collect.clear()
+
+            start_time = time.time()
+            dagbag.collect_dags(
+                dag_folder,
+                include_examples=configuration.getboolean('core', 'LOAD_EXAMPLES'))
+            event_collect_done.set()
+            event_next_collect.wait()
+
+            time.sleep(max(0, COLLECT_DAGS_INTERVAL - (time.time() - start_time)))
+        except SystemExit:
+            sys.exit(0)
+        except Exception:  # pylint: disable=broad-except
+            logging.warning('Dagbag loader dags collector errors.', exc_info=True)
+
+
+def _cleanup_dagbag(dagbag, current_keys):
+    """Delete keys in the dagbag but not in the latest collection."""
+    for k in ['dags', 'file_last_changed', 'import_errors']:
+        v = dagbag.__dict__[k]
+        for key_to_delete in set(v.keys()) - set(current_keys[k]):
+            del v[key_to_delete]
+        current_keys[k] = []
+
+
+def _receive_dagbag(dagbag, queue):
+    """A thread that receives updated dagbag."""
+    current_keys = defaultdict(list)
+    while True:
+        try:
+            collect_done, dagbag_update = queue.get()
+            for k, v in dagbag_update.items():
```

Comment:
logs are getting filled with this error: `AttributeError: 'tuple' object has no attribute 'items'`

Reply:
> logs are getting filled with this error: `AttributeError: 'tuple' object has no attribute 'items'`

My bad, it is a bug in code: line 100:
dagbag_update = (dagbag, previous_keys, collect_done) 
should be:
dagbag_update = _make_dagbag_update(dagbag, previous_keys, collect_done)